### ISSUE TITLE
Proposal: Updates to support Sitecore 9.x

### ISSUE THREAD

**Issue by sc-alistairdeneys** [sshushliapin/Sitecore.FakeDb#209]
I'd like to submit some contributions to add support for the 9.x Sitecore versions. Unfortunately there have been some breaking changes in the APIs which mean we'll need to handle each major Sitecore version differently.

Looking at the v2 branch, it would appear support for the older Sitecore versions has been removed, as would be necessary to get the library building. Therefore, I propose we increment the major version of FakeDB as necessary to add support for each of the Sitecore 9.x versions. Something like:

* FakeDB 1.x (current release) for Sitecore 7.5 - 8.2
* FakeDB 2.x (next immediate release) for Sitecore 9.0 - 9.1
* FakeDB 3.x for Sitecore 9.2 - 9.3

In addition to the versioning changes, I'd also like to make a few changes to make the project easier to work with:

* Migrate the projects over to NET SDK format (the format used for .net core projects) as it makes management of NuGet packages so much easier.
* Switch references over to using the official Sitecore NuGet feeds. All Sitecore 9.x versions are available on the feeds.

I intend to make these contributions myself.

Any concerns with this proposal?

**Comment by sshushliapin** [comment]
Hi @sc-alistairdeneys! Thanks for your input!

Versioning approach looks good to me.

Regarding the migration. I'd love to do so but it seems it requires some additional research. Now there is a custom [ConfigReader](https://github.com/sshushliapin/Sitecore.FakeDb/blob/master/src/Sitecore.FakeDb/Configuration/ConfigReader.cs) which performs some vital initial configurations. It's like an entry point to the lib! Sounds weird but it did the job for many years and without it nothing works :). And it stops working in SDK projects. Maybe I overlooked smth, but for now that's a stopper.

**Comment by sc-alistairdeneys** [comment]
My initial researched didn't show any issues; with an SDK format project everything was still working fine, including all the tests. I'll prep the PR and you can validate @sshushliapin.

**Comment by mikeedwards83** [comment]
To continue this conversation, will you be supporting 9.3?

**Comment by sc-alistairdeneys** [comment]
My initial reaction was to say "yes" for 9.3 support, but then I'm wondering whether we should sit back and see if it's necessary.

The abstractions in Sitecore have got a lot better, to the point that FakeDB shouldn't be required for unit testing on later versions of Sitecore. What do you think @sshushliapin?

I'll continue with adding the 9.x support (looks to be working locally, just need time to check fully) up until version 9.2.

**Comment by sshushliapin** [comment]
> The abstractions in Sitecore have got a lot better, to the point that FakeDB shouldn't be required for unit testing on later versions of Sitecore. What do you think @sshushliapin?

Yes, it is not required anymore. TBH, I haven't used FakeDb for the past two years...

I do not plan any further development but if it's possible to make it 9.2/9.3 compatible and it will work, then, of course, I'm ready to support this.

**Comment by MKargeti** [comment]
we are using Sitecore 9.2 and using Fake db in our project. any plans of releasing the compatible version of it or we have to re-write the test cases, please confirm

**Comment by sshushliapin** [comment]
According to #211 and #213, I'd expect all the 9.0-9.3 versions to be supported. Any specific scenario that does now work in your case? Otherwise, this issue should be closed.
